src/Analysis/Engine/Impl/DependencyResolution/PathResolverSnapshot.cs

```diff
@@ -0,0 +1,280 @@
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Microsoft.PythonTools.Analysis.Infrastructure;
+using Microsoft.PythonTools.Parsing;
+
+namespace Microsoft.PythonTools.Analysis.DependencyResolution {
+    internal partial struct PathResolverSnapshot {
+        private static readonly bool IgnoreCase = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+        private readonly PythonLanguageVersion _pythonLanguageVersion;
+        private readonly Node[] _roots;
+        public int Version { get; }
+
+        public PathResolverSnapshot(PythonLanguageVersion pythonLanguageVersion) {
+            _pythonLanguageVersion = pythonLanguageVersion;
+            _roots = Array.Empty<Node>();
+            Version = 0;
+        }
+
+        private PathResolverSnapshot(PythonLanguageVersion pythonLanguageVersion, Node[] roots, int version) {
+            _pythonLanguageVersion = pythonLanguageVersion;
+            _roots = roots;
+            Version = version;
+        }
+
+        public IAvailableImports GetAvailableImportsFromAbsolutePath(in string modulePath, in IEnumerable<string> path) {
+            if (!TryFindModule(modulePath, out var lastArc, out _, out _)) {
+                return default;
+            }
+
+            //TODO: handle implicit relative path for 2X, use TryAppendPackages(ref lastArc, path)
+            var firstArc = lastArc.GetFirstArc();
+            if (!TryAppendPackages(firstArc, path, out lastArc)) {
+                return default;
+            }
+
+            return GetAvailableImports(lastArc);
+        }
+
+        public IAvailableImports GetAvailableImportsFromRelativePath(in string modulePath, in int parentCount, in IEnumerable<string> relativePath) {
+            if (!TryFindModule(modulePath, out var lastArc, out _, out _)) {
+                // Module wasn't found in current snapshot
+                return default;
+            }
+
+            if (parentCount > lastArc.PathLength - 1) {
+                // Can't get outside of the root
+                return default;
+            }
+
+            var relativeParentArc = lastArc.GetPrevious(parentCount);
+            if (!TryAppendPackages(relativeParentArc, relativePath, out lastArc)) {
+                return default;
+            }
+            
+            return GetAvailableImports(lastArc);
+        }
+
+        private IAvailableImports GetAvailableImports(in Arc lastArc, in bool requireInitPy = true) {
```

Comment:
Do we need `requireInitPy`? The object appears to use specific language version that defines the rule.

Reply:
Yep, not required.